### PR TITLE
Improve test coverage and add edge case tests

### PR DESCRIPTION
**Pull request by mubarakcoded**
## Summary by CodeRabbit

* **Tests**
  * Expanded end-to-end and unit tests covering group permissions and creator validation, member management, metadata updates, message lifecycle and idempotency, multi-client synchronization and concurrent commit/race scenarios, welcome processing (size/format limits, recovery, and error paths), and key-package lifecycle validations.

* **New Features**
  * Added public test utilities: a multi-client test harness and a deterministic race-condition simulator to support integration and concurrency testing.

### CODE REVIEW

File: crates/mdk-core/src/groups.rs [hunk 1]
```diff
@@ -2543,4 +2543,225 @@ mod tests {
...
+        // This should fail because non-admin cannot add members
+        // Note: In practice, the non-admin wouldn't have the group loaded,
+        // but we're testing the permission check logic
+        let result = non_admin_mdk.add_members(&group_id, &[new_member_key_package]);
+
+        // The error should indicate permission denied or group not found
+        assert!(
+            result.is_err(),
+            "Non-admin should not be able to add members"
+        );
```
Comment (jgmontoya) [inline crates/mdk-core/src/groups.rs]: We should add an assertion that shows that the members did not change

File: crates/mdk-core/src/groups.rs [hunk 2]
```diff
@@ -2543,4 +2543,225 @@ mod tests {
...
+
+        // Try to have the non-admin update group name
+        let update = NostrGroupDataUpdate::new().name("Hacked Name".to_string());
+        let result = non_admin_mdk.update_group_data(&group_id, update);
+
+        // Should fail - non-admin cannot update group extensions
+        assert!(
+            result.is_err(),
+            "Non-admin should not be able to update group extensions"
+        );
```
Comment (jgmontoya) [inline crates/mdk-core/src/groups.rs]: same regarding the assertion but showing extensions did not change

**Review by jgmontoya — APPROVED** [review]
Nice!
